### PR TITLE
Convert all projects to SDK-style

### PR DESCRIPTION
**Pull request by wtgodbe**
Part of https://github.com/dotnet/corefx/projects/3

CC @weshaggard @joperezr @safern @eerhardt

### CODE REVIEW

**Comment by joperezr** [comment]
Did you require any buildtools changes with this? do you mind sharing those if you did?

File: src/Directory.Build.props [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildProjectDirectory), dir.props))\dir.props" />
+  <PropertyGroup>
+    <EnableDefaultItems>false</EnableDefaultItems>
+    <NoWarn>CS3021;$(NoWarn)</NoWarn>
+    <RestoreOutputPath>$(MSBuildThisFileDirectory)\System.Runtime\ref\</RestoreOutputPath>
```
Comment (weshaggard) [inline src/Directory.Build.props]: Why is this going to System.Runtime\ref?

File: src/Directory.Build.props [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    <IsSDKProject>true</IsSDKProject>
+  </PropertyGroup>
+  <PropertyGroup>
+    <CSharpCoreTargetsPath>$(MSBuildExtensionsPath32)\Roslyn\Microsoft.CSharp.Core.targets</CSharpCoreTargetsPath>
```
Comment (weshaggard) [inline src/Directory.Build.props]: Why do we need to explicitly set these here?

File: src/Directory.Build.targets [hunk 3]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project InitialTargets="AddSkipGetTargetFrameworkToProjectReferences">
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildProjectDirectory), dir.targets))\dir.targets" />
```
Comment (weshaggard) [inline src/Directory.Build.targets]: What are these targets for?

File: src/Common/perf/PerfRunner/PerfRunner.csproj [hunk 4]
```diff
@@ -1,6 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
-<Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
-  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (weshaggard) [inline src/Common/perf/PerfRunner/PerfRunner.csproj]: @eerhardt are you aware of any way we can avoid setting this string in all our projects? We are very likely going to need to move to another SDK as part of our arcade repo and it would be great if we could just change it in one place in the repo if possible.

**Comment by wtgodbe** [comment]
I have not made any yet - the current state leaves me with the following error:

> E:\code\corefx\Tools\dotnetcli\sdk\2.1.300-rc1-008673\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(197,5): error MSB4018: The "ResolvePackageAssets" task failed unexpectedly. [E:\code\corefx\src\System.Buffers\src\System.Buffers.csproj]
E:\code\corefx\Tools\dotnetcli\sdk\2.1.300-rc1-008673\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(197,5): error MSB4018: NuGet.Packaging.Core.PackagingException: Unable to find fallback package folder 'F:\git\corefx\Tools\dotnetcli\sdk\NuGetFallbackFolder'. [E:\code\corefx\src\System.Buffers\src\System.Buffers.csproj]

The fallback package folder location, `F:\git\corefx\Tools\dotnetcli\sdk\NuGetFallbackFolder`, is surprising to me. The base location (`F:\git`) was present in https://github.com/dotnet/corefx/pull/29831/commits/be0421d1844c525a74be4a20534a01d89976e65e#diff-6483fde6c45c90fdfe98ca72eadfa651R7, and was a remnant of changes I brought in from Jose's original branch, but I've since corrected it in https://github.com/dotnet/corefx/pull/29831/commits/f60fa382639a2aa8f500b75d11459759106d5ca0#diff-6483fde6c45c90fdfe98ca72eadfa651R7, and I don't see any other places where `F:\git` is hardcoded.

**Comment by weshaggard** [comment]
>Did you require any buildtools changes with this? do you mind sharing those if you did?

Yes I would expect a buildtools change as well.

File: src/Directory.Build.targets [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+  <Target Name="_CheckForInvalidConfigurationAndPlatform" />
+  <Target Name="_CheckForUnsupportedTargetFramework" />
```
Comment (weshaggard) [inline src/Directory.Build.targets]: Perhaps it is part of your buildtools changes but how are we blocking the import of the csharp targets?

File: src/Directory.Build.props (hunk 1 above)
Comment (wtgodbe) [reply src/Directory.Build.props]: This is a remnant of changes I brought in from Jose's original changeset - the location pointed to here needs a project.assets.json file, which I also added in this change. I'm not married to that particular location

File: src/Directory.Build.props (hunk 2 above)
Comment (wtgodbe) [reply src/Directory.Build.props]: This is also a remnant I took from Jose's original changeset - I haven't tried running a build without it yet.

File: src/Directory.Build.targets (hunk 5 above)
Comment (wtgodbe) [reply src/Directory.Build.targets]: I haven't explicitly done this yet, maybe a change I will have to implement in buildtools.

File: src/Directory.Build.props (hunk 1 above)
Comment (joperezr) [reply src/Directory.Build.props]: This is wrong, I had it like this as part of my changes only when I was iterating on some changes and wanted to verify that stuff was being restored correctly. This should be removed.

**Comment by safern** [comment]
> The fallback package folder location, F:\git\corefx\Tools\dotnetcli\sdk\NuGetFallbackFolder, is surprising to me. The base location (F:\git) was present in be0421d#diff-6483fde6c45c90fdfe98ca72eadfa651R7, and was a remnant of changes I brought in from Jose's original branch, but I've since corrected it in f60fa38#diff-6483fde6c45c90fdfe98ca72eadfa651R7, and I don't see any other places where F:\git is hardcoded.

Maybe because the project.assets.json path? 

https://github.com/dotnet/corefx/pull/29831/commits/f60fa382639a2aa8f500b75d11459759106d5ca0#diff-2a844428e665aad62dc165df96e249bdR18

File: src/Directory.Build.targets (hunk 3 above)
Comment (joperezr) [reply src/Directory.Build.targets]: This is just what imports the parent directory's dir.targets. We used to have this per project, and we now have it only once on the common file.

**Comment by wtgodbe** [comment]
@safern yep, that's definitely it... not sure how I missed that in the repo search. @joperezr should I remove both the `RestoreOutputPath` and `project.assets.json` altogether?

File: src/Directory.Build.targets (hunk 3 above)
Comment (safern) [reply src/Directory.Build.targets]: Should we instead of import dir.targets, rename dir.targets to Directory.Build.Targets?

Comment (wtgodbe) [reply src/Directory.Build.targets]: I would be fine with that, though it would probably confuse some devs at first

**Comment by joperezr** [comment]
To be honest, I don't remember entirely what that particular RestoreOutputPath was doing there, so I don't have a good answer to you. What I do know, is that the first project I converted and tried to get to compile succesfully was System.Runtime, which is probably related to the property. I would just remove it for now and try to build and see why/if it was needed.

**Comment by wtgodbe** [comment]
Updated to remove the project.assets.json file, now hitting the following error:

> E:\code\corefx\Tools\dotnetcli\sdk\2.1.300-rc1-008673\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(197,5): error : Assets file 'E:\code\corefx\src\System.Buffers\src\project.assets.json' not found. Run a NuGet package restore to generate this file. [E:\code\corefx\src\System.Buffers\src\System.Buffers.csproj]

File: src/Directory.Build.targets (hunk 3 above)
Comment (kasper3) [reply src/Directory.Build.targets]: Directory.Build.Targets is a new feature of msbuild that some repos under dotnet/aspnet have started to use already:
https://github.com/search?q=org:dotnet+filename:directory.build.targets
https://github.com/search?q=org:aspnet+filename:directory.build.targets

File: src/Common/perf/PerfRunner/PerfRunner.csproj (hunk 4 above)
Comment (kasper3) [reply src/Common/perf/PerfRunner/PerfRunner.csproj]: Based on https://github.com/Microsoft/msbuild/issues/1493, we can use `<Import>` semantics, because in `Project Sdk="$(prop)"`,  `prop` does not be evaluated, nor in `<Sdk Name="$(prop)">`. Alternatively, we can implement `ISdkResolver` when it is implemented in MSBuild. :)

File: src/Directory.Build.targets (hunk 3 above)
Comment (weshaggard) [reply src/Directory.Build.targets]: I do think we should rename dir.targets and dir.props to Directory.Build.props/targets.

Comment (wtgodbe) [reply src/Directory.Build.targets]: Do you think I should rename all dir.props/dir.targets to Directory.Build.props/targets? @joperezr is there documentation on how SDK projects automatically find files named as such? Will there be collisions if multiple files share that name across different parts of the directory tree?

Comment (eerhardt) [reply src/Directory.Build.targets]: > is there documentation on how SDK projects automatically find files named as such?

Check out https://docs.microsoft.com/en-us/visualstudio/msbuild/customize-your-build

File: src/Common/perf/PerfRunner/PerfRunner.csproj (hunk 4 above)
Comment (davkean) [reply src/Common/perf/PerfRunner/PerfRunner.csproj]: The error above indicates that you are not importing Microsoft.Common.targets and indirectly Microsoft.Managed.DesignTime.targets. You can verify this by doing a `msbuild /pp` on the project.

**Comment by wtgodbe** [comment]
Current state-

With the above error, it appears the error is happening because we're running the target `ResolvePackageAssets` against a project that hasn't had `restore` run it. It seems correct to me not to run `restore` on most CoreFx projects, since they don't restore any packages. `ResolvePackageAssets` runs because `ResolveLockFileReferences` depends on it, which runs because `ResolvePackageDependenciesForBuild` depends on it, which runs as a BeforeTarget of `AssignProjectConfiguration`. This all happens in Microsoft.PackageDependencyResolution.targets, an SDK targets file.

The only way to conditionally disable any of these targets is by setting `DesignTimeBuild = true`, and not having a Project Asset File present. I'm not sure what the implications of setting `DesignTimeBuild = true` are yet, though in trying it I hit another failure involving a missing contract (no .dll for the desired project in bin/ref).

As for `AssignProjectConfiguration`, it appears that runs because `ResolveReferences` depends on it, and it is conditioned on `'$(CurrentSolutionConfigurationContents)' != '' or '@(ProjectReference)'!=''`. This happens in `Microsoft.Common.CurrentVersion.targets`. Trying to decide what the best way to proceed from here is.

**Comment by davkean** [comment]
@wtgodbe Do not set DesignTimeBuild to true. This is intended for VS purposes only.

You should call Restore on the project - Restore is more than just packages, it's also contains information on indirect P2P.

File: src/Directory.Build.targets (hunk 3 above)
Comment (nguerrera) [reply src/Directory.Build.targets]: You can import the SDK in a central props/targets, but then the central props/targets can't be Directory.Build.*. Chicken and egg. Sdk imports common targets and common targets imports Directory.Build.*. 

If you use explicit dir.props/dir.targets, then you can do this: https://github.com/nguerrera/sdkimportrepro/commit/6d34578b41c4f4e68ec1aa3dace648923be75287

File: src/Common/perf/PerfRunner/PerfRunner.csproj (hunk 4 above)
Comment (nguerrera) [reply src/Common/perf/PerfRunner/PerfRunner.csproj]: You can import the SDK in a central props/targets, but then the central props/targets can't be `Directory.Build.*`. Chicken and egg. Sdk imports common targets and common targets imports `Directory.Build.*`

If you use explicit dir.props/dir.targets, then you can do this: nguerrera/sdkimportrepro@6d34578

Comment (nguerrera) [reply src/Common/perf/PerfRunner/PerfRunner.csproj]: Oh I see that @eerhardt  explicitly imported `Directory.Build.*` and then turned off the magic import. That should have worked too. My above commit with different name for explicit directory props/targets is working fine in VS. Let me try @eerhardt's approach in that sandbox.

Comment (nguerrera) [reply src/Common/perf/PerfRunner/PerfRunner.csproj]: https://github.com/nguerrera/sdkimportrepro/commit/ab90e6218bc67b55d4e3aa311881261849ea5cfb works fine in VS for me too. I cannot repro what @eerhardt is seeing with that approach.

**Comment by weshaggard** [comment]
@wtgodbe I agree with @davkean that we shouldn't hack this and set DesignTimeBuild.

@davkean for our corefx projects we have been disabling restore to avoid the extra cost for each project. We don't plan to use PackageReference's and there will be limited ProjectReferences in our repo, as we plan to override the target framework to reference a folder of refs that we build in our repo. Do you think the performance is good enough to try and enable restore for all our projects even though we expect it to be a no-op for almost all? Or should we try a disable restore in our common targets?

**Comment by davkean** [comment]
Whether performance is good is the eye of the beholder; you should try using using implicit restore via MSBuild `/restore` switch or via `dotnet build` first. We've worked really hard to improve performance for this case. In VS, we're going to restore regardless - and if you want consistent behavior based on whether you only use command-line or used VS -> switched to command-line, I would just always restore.

**Comment by weshaggard** [comment]
Thanks @davkean we definitely want to try and have consistency between command line and VS. We will try the default restore way first and measure before making a decision to disable it or not. 

>In VS, we're going to restore regardless 

That isn't controllable via an msbuild property from the project?

**Comment by davkean** [comment]
You can turn if via NuGet.config to prevent it from occurring automatically (say when you modify a project file or switch branches) and/or during build - but I see no reason why you would turn this on, unless it was causing issues, at which point we'd want bugs to fix said issues.

Using Visual Studio is much nicer if you just let us do our thing to make the project in the state we think it should be.

**Comment by davkean** [comment]
> and if you want consistent behavior based on whether you only use command-line or used VS -> switched to command-line, I would just always restore.
On this, by consistent behavior, I mean you might get different build errors based on whether you had previously restored using VS or not. For example, let's say you have A -> B -> C. For normal SDK-based projects, 'A' can use 'C'. If you turn off restore in command-line but not for VS - you're not going to different build error based on whether you've ever opened in VS or not.

**Comment by kasper3** [comment]
Probably we don't need to explicitly run `dotnet restore` or `dotnet msbuild /restore`

"`cd` into any corefx project directory and run `dotnet build` for default behavior (e.g. build) or `dotnet build /p:MyProp=MyVal /t:MySpecialTarget`"
is much cooler and popular than `dotnet msbuild`.

File: src/Directory.Build.targets (hunk 3 above)
Comment (4creators) [reply src/Directory.Build.targets]: dir.props and dir.targets are maltilayered in many repos (they are present at multiple directory levels and recursively each one includes higher level projects). Directory.Build.props/.targets are included once - after first include higher level ones will be skipped. This may cause some problems unless recursive include logic from dirs.props is applied to Directory.Build.* files as well.

Comment (eerhardt) [reply src/Directory.Build.targets]: Yes, this is expected with Directory.Build.*.  It gets imported once, and then if there are parent props/targets files, they need to be included up the chain.  This allows props/target authors to get the ordering right, which is important especially in props files.

We do this in the dotnet/machinelearning repo: https://github.com/dotnet/machinelearning/blob/master/src/Directory.Build.props#L3

File: src/Common/perf/PerfRunner/PerfRunner.csproj (hunk 4 above)
Comment (eerhardt) [reply src/Common/perf/PerfRunner/PerfRunner.csproj]: I tried it on today's daily dogfood build, and am no longer seeing the error.  The project loads and builds correctly in VS.

**Comment by weshaggard** [comment]
>Probably we don't need to explicitly run dotnet restore or dotnet msbuild /restore

I agree for individual project builds we will just do dotnet build, but depending on the performance impact across the building of all projects we may do an upfront restore of all projects before we do a build, we will have to see the impact first.

**Comment by wtgodbe** [comment]
running `dotnet.exe msbuild /t:restore System.Buffers.csproj` gives:

> E:\code\corefx\Tools\dotnetcli\sdk\2.1.300-rc1-008673\NuGet.targets(114,5): error : Invalid framework identifier ''. [E:\code\corefx\src\System.Buffers\src\System.Buffers.csproj]

Here's the file in question: https://github.com/dotnet/corefx/blob/master/src/System.Buffers/src/System.Buffers.csproj

@davkean what's the right property to be setting here?

**Comment by kasper3** [comment]
@natemcmaster, @nguerrera, could anyone from aspnet/universe or dotnet/sdk team up to help on this PR <3

**Comment by eerhardt** [comment]
I'll take a look.  I'm getting a `/bl` now.

**Comment by wtgodbe** [comment]
Per Eric, we need to have values set for the properties in the following block in NuGet.Targets:

`<GetProjectTargetFrameworksTask
      Condition=" '$(RestoreProjectStyle)' != 'ProjectJson' "
      ProjectPath="$(MSBuildProjectFullPath)"
      TargetFrameworks="$(TargetFrameworks)"
      TargetFramework="$(TargetFramework)"
      TargetFrameworkMoniker="$(TargetFrameworkMoniker)"
      TargetPlatformIdentifier="$(TargetPlatformIdentifier)"
      TargetPlatformVersion="$(TargetPlatformVersion)"
      TargetPlatformMinVersion="$(TargetPlatformMinVersion)"> `

We should be setting `TargetFrameworks` to the full list of TFMs we support per package. Currently trying to figure out how to determine that list.

**Comment by weshaggard** [comment]
@wtgodbe have you actually made any buildtools changes locally yet? If not then you aren't going to have any luck getting our projects to build that way.

**Comment by wtgodbe** [comment]
I'm not sure what needs to change, as of right now I'm still erroring at `GetProjectTargetFrameworksTask` as mentioned above. I was under the impression CoreFx needed to set up the right properties for that task, should I be trying to gather this information in BuildTools instead? @joperezr is this something you hit as well?

**Comment by joperezr** [comment]
> is this something you hit as well?

I don't remember for sure what was I hitting, but I was hitting a similar thing when trying to restore since the project.assets.json file required TargetFramework to be set. For this reason, I generated one assets file and then shared it accross all projects.

File: Directory.Build.props [hunk 6]
```diff
@@ -1,6 +1,10 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
-  <Import Condition="Exists('..\dir.props')" Project="..\dir.props" />
+
+  <PropertyGroup>
+    <!-- Don't need Microsoft.Common.props to import Directory.Build.props again -->
+    <ImportDirectoryBuildProps>false</ImportDirectoryBuildProps>
```
Comment (weshaggard) [inline Directory.Build.props]: Why is this necessary?

Comment (eerhardt) [reply Directory.Build.props]: Should I make the comment better here?

Comment (weshaggard) [reply Directory.Build.props]: The comment is definitely added confusion to me as I was expecting this to automatically be imported. Is there any way we can just let the SDK import these in those project types as well? Or are they too much work right now to convert to SDK?

Comment (eerhardt) [reply Directory.Build.props]: The way `Directory.Build.props|targets` files get automatically imported is by someone importing `Microsoft.Common.props|targets`.  If your project isn't an "SDK-style" project, then somewhere in your project, you need an explicit `<Import>` statement in order to get the `Microsoft.Common.props|targets` imported.

So in order to do that, we would either need to:
1. Make depproj and pkgproj (basically ALL our project files in corefx) be some sort of "SDK-style" projects. And in that "SDK", it would need to import `Microsoft.Common.props|targets`.
2. Explicitly `Import` a different `.props|.targets` set of files at the top and bottom of the .depproj and .pkgproj files.  Then that other set of `.props|.targets` files would import `Microsoft.Common.props|targets` at the right place, which would then import the `Directory.Build.props|targets` correctly.

I haven't tried (1) above, yet.  You said @joperezr was working on the pkgproj projects, so I didn't want to do any work there.  Also, it feels like we should try to make incremental progress here - get the .csproj's converted first, then worry about .depproj and .pkgproj.

Comment (weshaggard) [reply Directory.Build.props]: For now I'm fine if you could please update the comment to call out this is about dealing with the non-sdk projects.

Comment (eerhardt) [reply Directory.Build.props]: Done.

**Comment by wtgodbe** [comment]
Error dump for building in VS (After making changes):

> Assets file 'E:\code\corefx\src\System.Collections\ref/project.assets.json' not found. Run a NuGet package restore to generate this file.

> The "FindBestConfiguration" task failed unexpectedly - could not find a value for 'TargetGroup' and no default value exists (VS Crashes when I try to copy this message, so this is a rough transcription of the error)

Trying to figure out what's going on now.

**Comment by davkean** [comment]
@wtgodbe The first is likely caused by the second if it prevents us from doing a [design-time build](https://github.com/dotnet/project-system/blob/master/docs/design-time-builds.md). See https://github.com/dotnet/project-system/blob/master/docs/design-time-builds.md#visual-studio-2017-version-154-or-later for grabbing the results of the builds to figure it out. FindBestConfiguration doesn't sound like a built-in target, so I would first start with it.

**Comment by eerhardt** [comment]
@wtgodbe @weshaggard @joperezr - I think this may turn green.  It would be nice if someone other than me pulled the branch, and tested out some dev scenarios in VS.  To do that, you also need [this change to dotnet/buildtools](https://github.com/eerhardt/buildtools/commit/c44fa3a870de97a3c591e7587cdaa3ea947004ad) (which I'll make a PR for tomorrow).

**Comment by eerhardt** [comment]
Test Windows x64 Debug Build

**Comment by wtgodbe** [comment]
@eerhardt I'm pulling the branch for testing now

**Comment by wtgodbe** [comment]
@eerhardt I tried building from the command line, then opening System.Buffers.sln & building src\System.Buffers.csrpoj from the command line in VS2017. I got the following error:

> A project with an Output Type of Class Library cannot be started directly. In order to debug this project, add an executable project to this solution which references the library project. Set the executable project as the startup project.

It could be that there's something wrong with my configuration.

**Comment by eerhardt** [comment]
You got that error while *building*?  That seems surprising.

Note: I assume you pulled the buildtools change locally, built it, and used that buildtools for corefx?

**Comment by wtgodbe** [comment]
@eerhardt yeah, I did the following things:

1. clone & build https://github.com/eerhardt/buildtools/commit/c44fa3a870de97a3c591e7587cdaa3ea947004ad
2. clone this branch
3. reference the built buildtools in BuildToolsVersion.txt/dependencies.props/init-tools.cmd
4. run build.cmd & confirm that the right Buildtools version is in my Tools dir
5. Open System.Buffers.Sln in VS2017 & hit 'build'

**Comment by wtgodbe** [comment]
Never mind, I was building wrong. Build looks good in my (older) version of VS2017. I'm running a code coverage test run right now - it's taking a while, but seems to be working so far.

**Comment by wtgodbe** [comment]
I got the following errors while running `build-test -coverage` on Windows:

> E:\code\corefx\Tools\tests.targets(522,5): error : One or more tests failed while running tests from 'System.Console.Tests' please check E:\code\corefx\bin/tests/System.Console.Tests/netcoreapp-Windows_NT-Debug-x64/testResults.xml for details! [E:\code\corefx\src\System.Console\tests\System.Console.Tests.csproj]
E:\code\corefx\Tools\tests.targets(522,5): error : One or more tests failed while running tests from 'System.Globalization.CalendarsWithConfigSwitch.Tests' please check E:\code\corefx\bin/tests/System.Globalization.CalendarsWithConfigSwitch.Tests/netcoreapp-Windows_NT-Debug-x64/testResults.xml for details! [E:\code\corefx\src\System.Globalization.Calendars\tests\CalendarTestWithConfigSwitch\System.Globalization.CalendarsWithConfigSwitch.Tests.csproj]
E:\code\corefx\Tools\tests.targets(522,5): error : One or more tests failed while running tests from 'Invariant.Tests' please check E:\code\corefx\bin/tests/Invariant.Tests/netcoreapp-Windows_NT-Debug-x64/testResults.xml for details! [E:\code\corefx\src\System.Globalization\tests\Invariant\Invariant.Tests.csproj]
E:\code\corefx\Tools\tests.targets(522,5): error : One or more tests failed while running tests from 'System.IO.IsolatedStorage.Tests' please check E:\code\corefx\bin/tests/System.IO.IsolatedStorage.Tests/netcoreapp-Windows_NT-Debug-x64/testResults.xml for details! [E:\code\corefx\src\System.IO.IsolatedStorage\tests\System.IO.IsolatedStorage.Tests.csproj]

I did get a coverage report, though. I'll compare that to the result of just running build-test to see if the errors are specific to the coverage run.

**Comment by wtgodbe** [comment]
This is what I get running a normal test build:

> E:\code\corefx\dir.traversal.targets(77,5): error : (No message specified) [E:\code\corefx\src\tests.builds]

Build FAILED.

E:\code\corefx\Tools\tests.targets(514,5): warning :    System.Console.Tests  Total: 120, Errors: 0, Failed: 11, Skipped: 2, Time: 50.245s [E:\code\corefx\src\System.Console\tests\System.Console.Tests.csproj]
E:\code\corefx\Tools\tests.targets(514,5): warning MSB3073: The command "E:\code\corefx\bin/tests/System.Console.Tests/netcoreapp-Windows_NT-Debug-x64//RunTests.cmd E:\code\corefx\bin/testhost/netcoreapp-Windows_NT-Debug-x64/" exited with code 1. [E:\code\corefx\src\System.Console\tests\System.Console.Tests.csproj]
E:\code\corefx\Tools\tests.targets(522,5): error : One or more tests failed while running tests from 'System.Console.Tests' please check E:\code\corefx\bin/tests/System.Console.Tests/netcoreapp-Windows_NT-Debug-x64/testResults.xml for details! [E:\code\corefx\src\System.Console\tests\System.Console.Tests.csproj]
E:\code\corefx\dir.traversal.targets(77,5): error : (No message specified) [E:\code\corefx\src\tests.builds]

**Comment by eerhardt** [comment]
I'm removing the WIP on this PR.  I believe this is ready for review and to be merged.

All the legs should be passing - I will fix them if they aren't.
I've tested out the VS experience. Overall, things work very well. However, there are 2 bugs I've run into:
1. VS consumes a lot of CPU in the background https://github.com/dotnet/project-system/issues/3670#issuecomment-402883217
2. The editor doesn't pick up the new properties after switching configurations https://github.com/dotnet/project-system/issues/2733#issuecomment-402885190  (unloading and reloading the project fixes it)

@weshaggard @wtgodbe - please take a look. Since this change is so big, it will probably be easiest to review this change commit-by-commit, and possibly skipping the merge commits.  Unless you want to pull the change locally and review it that way.

File: Directory.Build.props [hunk 7]
```diff
@@ -312,6 +319,15 @@
     <IncludeDllSafeSearchPathAttribute Condition="'$(TargetGroup)' == 'netstandard1.0'">false</IncludeDllSafeSearchPathAttribute>
   </PropertyGroup>
 
+  <!-- Set up default paths for reference assemblies -->
+  <PropertyGroup Condition="'$(IsReferenceAssembly)'=='true'">
+    <!-- Create a common root output directory for all reference assemblies -->
+    <ReferenceAssemblyOutputPath Condition="'$(ReferenceAssemblyOutputPath)' == ''">$(BaseOutputPath)ref/</ReferenceAssemblyOutputPath>
```
Comment (weshaggard) [inline Directory.Build.props]: Why did these get moved here instead of using the ones in BuildTools? https://github.com/dotnet/buildtools/blob/16a6d1f39cf0a572041d51bd318913548ba41b34/src/Microsoft.DotNet.Build.Tasks/PackageFiles/ReferenceAssemblies.targets#L15

File: src/System.Net.NetworkInformation/src/System.Net.NetworkInformation.csproj [hunk 8]
```diff
@@ -5,7 +5,7 @@
     <ProjectGuid>{3CA89D6C-F8D1-4813-9775-F8D249686E31}</ProjectGuid>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <ILLinkClearInitLocals>true</ILLinkClearInitLocals>
-    <Configurations>netcoreapp-Linux-Debug;netcoreapp-Linux-Release;netcoreapp-OSX-Debug;netcoreapp-OSX-Release;netcoreapp-Windows_NT-Debug;netcoreapp-Windows_NT-Release;uap-Windows_NT-Debug;uap-Windows_NT-Release</Configurations>
+    <Configurations>netcoreapp-Linux-Debug;netcoreapp-Linux-Release;netcoreapp-OSX-Debug;netcoreapp-OSX-Release;netcoreapp-Unix-Debug;netcoreapp-Unix-Release;netcoreapp-Windows_NT-Debug;netcoreapp-Windows_NT-Release;uap-Windows_NT-Debug;uap-Windows_NT-Release</Configurations>
```
Comment (weshaggard) [inline src/System.Net.NetworkInformation/src/System.Net.NetworkInformation.csproj]: Do you know if these have to be static and directly in the project? I'm asking mainly to try and determine if we can combine the Debug/Release aspects with our existing BuildConfigurations property to eliminate the duplication.

File: Directory.Build.props (hunk 7 above)
Comment (eerhardt) [reply Directory.Build.props]: I’m not entirely sure why this changed. I asked @wtgodbe and he said that change was included with @joperezr’s initial changes. I can try reverting it to see what happens.

File: src/System.Net.NetworkInformation/src/System.Net.NetworkInformation.csproj (hunk 8 above)
Comment (eerhardt) [reply src/System.Net.NetworkInformation/src/System.Net.NetworkInformation.csproj]: No I don’t know for sure. @davkean or @Pilchie - do you know off the top of your head?

Comment (Pilchie) [reply src/System.Net.NetworkInformation/src/System.Net.NetworkInformation.csproj]: I believe it has to be known at evaluation time (i.e. Not the result of a target), but doesn't have to be directly in the project.  @davkean would know for sure.

File: Directory.Build.props (hunk 7 above)
Comment (weshaggard) [reply Directory.Build.props]: I'd like to understand if it is needed here still.

**Comment by eerhardt** [comment]
@dotnet-bot test this please

**Comment by eerhardt** [comment]
@dotnet-bot test this please

**Comment by eerhardt** [comment]
Test Windows x64 Debug Build
